pinot-core/src/main/java/org/apache/pinot/core/transport/DirectOOMHandler.java

```diff
@@ -0,0 +1,95 @@
+ * large data tables from servers concurrently. We want to close all channels to servers to proactively release
+ * the direct memory, because the execution of netty threads can all block in allocating direct memory, in which case
+ * no one will reach channelRead0.
+ */
+public class DirectOOMHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataTableHandler.class);
+  private static final AtomicBoolean DIRECT_OOM_SHUTTING_DOWN = new AtomicBoolean(false);
+  private final QueryRouter _queryRouter;
+  private final ServerRoutingInstance _serverRoutingInstance;
+  private final ConcurrentHashMap<ServerRoutingInstance, ServerChannels.ServerChannel> _serverToChannelMap;
+  private volatile boolean _silentShutDown = false;
+
+  public DirectOOMHandler(QueryRouter queryRouter, ServerRoutingInstance serverRoutingInstance,
+      ConcurrentHashMap<ServerRoutingInstance, ServerChannels.ServerChannel> serverToChannelMap) {
+    _queryRouter = queryRouter;
+    _serverRoutingInstance = serverRoutingInstance;
+    _serverToChannelMap = serverToChannelMap;
+  }
+
+  public void setSilentShutDown() {
+    _silentShutDown = true;
+  }
+
+  @Override
+  public void channelInactive(ChannelHandlerContext ctx) {
+    // if we are shutting down channels due to direct memory OOM, we short circuit the channel inactive
+    if (_silentShutDown) {
+      return;
+    }
+    ctx.fireChannelInactive();
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+    // catch direct memory oom here
+    if (cause instanceof OutOfMemoryError && cause.getMessage().contains("Direct buffer")) {
+      // only one thread can get here and do the shutdown
+      if (DIRECT_OOM_SHUTTING_DOWN.compareAndSet(false, true)) {
+        try {
+          LOGGER.error("Closing ALL channels to servers, as we are running out of direct memory "
+              + "while receiving response from {}", _serverRoutingInstance, cause);
+          // close all channels to servers
+          _serverToChannelMap.keySet().forEach(serverRoutingInstance -> {
+            ServerChannels.ServerChannel removed = _serverToChannelMap.remove(serverRoutingInstance);
+            removed.closeChannel();
+            removed.setSilentShutdown();
+          });
+          _queryRouter.markServerDown(_serverRoutingInstance,
+              new QueryCancelledException("Query cancelled as broker is out of direct memory"));
+          BrokerMetrics.get().addMeteredGlobalValue(BrokerMeter.DIRECT_MEMORY_OOM, 1L);
+        } catch (Exception e) {
+          LOGGER.error("Caught exception while handling direct memory OOM", e);
+        } finally {
+          DIRECT_OOM_SHUTTING_DOWN.set(false);
+        }
+      } else {
+        LOGGER.warn("Caught direct memory OOM, but another thread is already handling it", cause);
```

Comment:
Can we do this right after line 70 (as soon as we enter the if block) to ensure it does get emitted even in the event of any failure in the above code while closing channels ?

Reply:
In the event of exceptions in handling failures, two other logs will be emitted:
` LOGGER.error("Closing ALL channels to servers, as we are running out of direct memory "
              + "while receiving response from {}", _serverRoutingInstance, cause);`
and
` LOGGER.error("Caught exception while handling direct memory OOM", e);`

I have observed npe being recorded clearly with this; that's also why I added null checks in removed.closeChannel().

Reply:
I have moved the metric sending befor L70, which would give better insights